cloud/services/managedclusters/managedclusters.go

```diff
@@ -123,50 +125,36 @@ func (s *Service) Reconcile(ctx context.Context, spec interface{}) error {
 			},
 			AgentPoolProfiles: &[]containerservice.ManagedClusterAgentPoolProfile{},
 			NetworkProfile: &containerservice.NetworkProfileType{
-				NetworkPlugin:   containerservice.Azure,
-				LoadBalancerSku: containerservice.Standard,
+				NetworkPlugin:   containerservice.NetworkPlugin(*managedClusterSpec.NetworkPlugin),
+				LoadBalancerSku: containerservice.LoadBalancerSku(*managedClusterSpec.LoadBalancerSKU),
+				NetworkPolicy:   containerservice.NetworkPolicy(*managedClusterSpec.NetworkPolicy),
 			},
 		},
 	}
 
-	if managedClusterSpec.NetworkPlugin != nil {
-		properties.NetworkProfile.NetworkPlugin = containerservice.NetworkPlugin(*managedClusterSpec.NetworkPlugin)
-	}
-
 	if managedClusterSpec.PodCIDR != "" {
 		properties.NetworkProfile.PodCidr = &managedClusterSpec.PodCIDR
 	}
 
 	if managedClusterSpec.ServiceCIDR != "" {
-		properties.NetworkProfile.ServiceCidr = &managedClusterSpec.ServiceCIDR
-		ip, _, err := net.ParseCIDR(managedClusterSpec.ServiceCIDR)
-		if err != nil {
-			return fmt.Errorf("failed to parse service cidr: %w", err)
-		}
-		// HACK: set the last octet of the IP to .10
-		// This ensures the dns IP is valid in the service cidr without forcing the user
-		// to specify it in both the Capi cluster and the Azure control plane.
-		// https://golang.org/src/net/ip.go#L48
-		ip[15] = byte(10)
-		dnsIP := ip.String()
-		properties.NetworkProfile.DNSServiceIP = &dnsIP
-
-	}
-
-	if managedClusterSpec.NetworkPolicy != nil {
-		if strings.EqualFold(*managedClusterSpec.NetworkPolicy, "Azure") {
-			properties.NetworkProfile.NetworkPolicy = containerservice.NetworkPolicyAzure
-		} else if strings.EqualFold(*managedClusterSpec.NetworkPolicy, "Calico") {
-			properties.NetworkProfile.NetworkPolicy = containerservice.NetworkPolicyCalico
+		if managedClusterSpec.DNSServiceIP == nil {
+			properties.NetworkProfile.ServiceCidr = &managedClusterSpec.ServiceCIDR
+			ip, _, err := net.ParseCIDR(managedClusterSpec.ServiceCIDR)
+			if err != nil {
+				return fmt.Errorf("failed to parse service cidr: %w", err)
+			}
+			// HACK: set the last octet of the IP to .10
+			// This ensures the dns IP is valid in the service cidr without forcing the user
+			// to specify it in both the Capi cluster and the Azure control plane.
+			// https://golang.org/src/net/ip.go#L48
+			ip[15] = byte(10)
+			dnsIP := ip.String()
+			properties.NetworkProfile.DNSServiceIP = &dnsIP
 		} else {
-			return fmt.Errorf("invalid network policy: '%s'. Allowed options are 'calico' and 'azure'", *managedClusterSpec.NetworkPolicy)
+			properties.NetworkProfile.DNSServiceIP = managedClusterSpec.DNSServiceIP
```

Comment:
if we have both serviceCIDR and DNSServiceIP defined, please validate that DNSServiceIP is in the block defined by serviceCIDR and return an error if it's not.

https://golang.org/pkg/net/#IPNet.Contains

Reply:
Initially I wanted to include that as one of the validations, the only problem is that the serviceCIDR field resides in the owner cluster object and the capi-webhook serviceAccount doesn't have permissions to get cluster objects.  I believe we ended up deciding to just validate that the DNSServiceIp field represents a valid ip.  Thoughts?

Reply:
we don't need to do it in the webhook, but we should still fail fast in the reconciler itself (where we have both objects, as well as RBAC rights)

Reply:
Good catch, that makes sense :smile:

Reply:
still missing this one?

Reply:
nvm, found it 👍 https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/862/files#diff-78feb4e2f24becff4de1a0bc4348e52bR172